src/python_bindings/mod.rs

```diff
@@ -1,98 +1,165 @@
+    #[new]
+    fn __new__(index: PyIndex) -> Self {
+        PyGuide {
+            state: index.get_initial_state(),
+            index,
         }
     }
-}
 
-// FIXME: could be costly, confirm if FSMInfo will actually be part of the interface
-impl From<&PyFSMInfo> for FSMInfo {
-    fn from(fsm_info: &PyFSMInfo) -> Self {
-        FSMInfo {
-            initial: fsm_info.initial,
-            finals: fsm_info.finals.clone(),
-            transitions: fsm_info.transitions.clone(),
-            alphabet_anything_value: fsm_info.alphabet_anything_value,
-            alphabet_symbol_mapping: fsm_info.alphabet_symbol_mapping.clone(),
+    fn get_start_tokens(&self) -> PyResult<Vec<TokenId>> {
+        self.index
+            .get_allowed_tokens(self.index.get_initial_state())
+            .ok_or(PyErr::new::<PyValueError, _>(
+                "Initial state is not in allowed tokens",
+            ))
+    }
+
+    fn read_next_token(&mut self, token_id: TokenId) -> PyResult<Vec<TokenId>> {
+        match self.index.get_next_state(self.state, token_id) {
+            Some(new_state) => {
+                self.state = new_state;
+                self.index
+                    .get_allowed_tokens(new_state)
+                    .ok_or(PyErr::new::<PyValueError, _>(format!(
+                        "No token ids found for the next state: {new_state}"
+                    )))
+            }
+            None => Err(PyErr::new::<PyValueError, _>(format!(
+                "No next state found for the current state: {} with token ID: {token_id}",
+                self.state
+            ))),
         }
     }
-}
 
-#[pymethods]
-impl PyFSMInfo {
-    #[new]
-    fn new(
-        initial: State,
-        finals: FxHashSet<State>,
-        transitions: FxHashMap<(State, TransitionKey), State>,
-        alphabet_anything_value: TransitionKey,
-        alphabet_symbol_mapping: FxHashMap<String, TransitionKey>,
-    ) -> Self {
-        FSMInfo::new(
-            initial,
-            finals,
-            transitions,
-            alphabet_anything_value,
-            alphabet_symbol_mapping,
+    fn is_finished(&self) -> bool {
+        self.index.is_final_state(self.state)
+    }
+
+    fn __repr__(&self) -> String {
+        format!(
+            "Guide object with the state={:#?} and {:#?}",
+            self.state, self.index
         )
-        .into()
+    }
+
+    fn __str__(&self) -> String {
+        format!(
+            "Guide object with the state={} and {}",
+            self.state, self.index.0
+        )
+    }
+
+    fn __eq__(&self, other: &PyGuide) -> bool {
+        self == other
+    }
+
+    fn __reduce__(&self) -> PyResult<(PyObject, (Vec<u8>,))> {
+        Python::with_gil(|py| {
+            let cls = PyModule::import_bound(py, "outlines_core.fsm.outlines_core_rs")?
+                .getattr("Guide")?;
+            let binary_data: Vec<u8> =
+                bincode::encode_to_vec(self, config::standard()).map_err(|e| {
```

Comment:
This is fine for now, but we might want to consider versioning at one point. If new fields are added or representation changes, we won't be able to pickle old guides. This could prevent users from updating the version to avoid breaking existing deployments.

Reply:
There is a serialization investigation issue opened to answer this among other things: https://github.com/dottxt-ai/outlines-core/issues/101